obelisk/cpp/obelisk_cpp/include/obelisk_mujoco_sim_robot.h

```diff
@@ -410,81 +411,374 @@ namespace obelisk {
+                                                        << sensor_names.at(i));
+                            }
+                        } else {
+                            RCLCPP_ERROR_STREAM(
+                                this->get_logger(),
+                                "Sensor " << sensor_names.at(i)
+                                          << " is not associated with a valid Mujoco sensor type! Current sensor type: "
+                                          << mj_sensor_types.at(i));
+                        }
+
+                        GetTimeFromSim(msg.header.stamp.sec, msg.header.stamp.nanosec);
+                    }
+                    publisher->publish(msg);
+                };
+
+                RCLCPP_INFO_STREAM(this->get_logger(), "Timer callback created for a IMU!");
+                return cb;
+            } else if constexpr (std::is_same<MessageT, obelisk_sensor_msgs::msg::ObkFramePose>::value) {
+                // ------------------------------------------ //
+                // ---------- Frame Pose call back ---------- //
+                // ------------------------------------------ //
+                auto cb = [publisher, sensor_names, mj_sensor_types, this]() {
+                    obelisk_sensor_msgs::msg::ObkFramePose msg;
+
+                    bool has_framepos  = false;
+                    bool has_framequat = false;
+
+                    std::lock_guard<std::mutex> lock(sensor_data_mut_);
+                    for (size_t i = 0; i < sensor_names.size(); i++) {
+                        int sensor_id = mj_name2id(this->model_, mjOBJ_SENSOR, sensor_names.at(i).c_str());
+                        if (sensor_id == -1) {
+                            throw std::runtime_error("Sensor not found in Mujoco! Make sure your XML has the sensor.");
+                        }
+
+                        // *** Note *** We only use the frame name and relative frame for the framepos. Be sure that
+                        // framepos and framequat have the same object/frame. Framequat is always given in global
+                        // coordinates
+                        int sensor_addr = this->model_->sensor_adr[sensor_id];
+                        if (mj_sensor_types.at(i) == "framepos") {
+                            if (!has_framepos) {
+                                msg.position.x = this->data_->sensordata[sensor_addr];
+                                msg.position.y = this->data_->sensordata[sensor_addr + 1];
+                                msg.position.z = this->data_->sensordata[sensor_addr + 2];
+
+                                // Framepos sensors can be mounted to different objects
+                                int obj_type = this->model_->sensor_objtype[sensor_id];
+                                if (obj_type == mjOBJ_SITE) {
+                                    int site_id = this->model_->sensor_objid[sensor_id];
```

Comment:
the x_id variables here are all defined identically. you can probably define it right under obj_type just once and call it obj_id, passing it as the index in every if block

Reply:
i resolved this